### PR TITLE
vpn: kill orphan server processes on startup

### PR DESCRIPTION
**Pull request by wrboyce**
Change-type: patch
Signed-off-by: Will Boyce \<will@balena.io\>

---
##### Contributor checklist

- [X] Build output has been rebuilt and tested
- [ ] Introduces security considerations
- [ ] Tests are included
- [ ] Documentation is added or changed
- [ ] Affects the development, build or deployment processes of the component

##### Reviewer Guidelines
- When submitting a review, please pick:
  - '*Approve*' if this change would be acceptable in the codebase (even if there are minor or cosmetic tweaks that could be improved).
  - '*Request Changes*' if this change would not be acceptable in our codebase (e.g. bugs, changes that will make development harder in future, security/performance issues, etc).
  - '*Comment*' if you don't feel you have enough information to decide either way (e.g. if you have major questions, or you don't understand the context of the change sufficiently to fully review yourself, but want to make a comment)
---

### CODE REVIEW

File: src/utils/openvpn.ts [hunk 1]
```diff
@@ -267,7 +272,18 @@ export class VpnManager extends EventEmitter {
 		return emitter;
 	};
 
+	private killOrphan = () => {
+		if (fs.existsSync(this.pidFile)) {
+			const pid = fs
+				.readFileSync(this.pidFile)
+				.toString()
```
Comment (Page-) [inline src/utils/openvpn.ts]:
```suggestion
				.readFileSync(this.pidFile. 'utf8')
```

File: src/utils/openvpn.ts [hunk 2]
```diff
@@ -267,26 +278,53 @@ export class VpnManager extends EventEmitter {
...
+		}
+		const pid = (await fs.readFile(this.pidFile, 'utf8')).replace(/\D/g, '');
+		const signals = { PING: 0, KILL: 9 } as const;
+		const kill = (signal?: ValueOf<typeof signals>) =>
+			spawn('/bin/kill', signal != null ? [`-${signal}`, pid] : [pid]);
+
+		return new Promise(resolve => {
+			const timeout = Date.now() + 5 * 1000;
+			const waitForDeath = (code?: number) => {
+				if (code !== 0) {
```
Comment (Page-) [inline src/utils/openvpn.ts]: If it doesn't equal 0?

File: src/utils/openvpn.ts [hunk 3]
```diff
@@ -19,11 +19,18 @@ import * as Bluebird from 'bluebird';
...
 import * as es from 'event-stream';
 import { EventEmitter } from 'eventemitter3';
+import * as _fs from 'fs';
 import * as net from 'net';
 import VpnConnector = require('telnet-openvpn');
+import { promisify } from 'util';
 
 import { Netmask } from './netmask';
 
+const fs = {
```
Comment (Page-) [inline src/utils/openvpn.ts]: You can use fs.promises btw

File: src/utils/openvpn.ts [hunk 4]
```diff
@@ -267,26 +278,53 @@ export class VpnManager extends EventEmitter {
...
+		if (!(await fs.exists(this.pidFile))) {
+			return;
+		}
+		const pid = (await fs.readFile(this.pidFile, 'utf8')).replace(/\D/g, '');
+		const signals = { PING: 0, KILL: 9 } as const;
+		const kill = (signal?: ValueOf<typeof signals>) =>
+			spawn('/bin/kill', signal != null ? [`-${signal}`, pid] : [pid]);
+
+		return new Promise(resolve => {
+			const timeout = Date.now() + 5 * 1000;
```
Comment (Page-) [inline src/utils/openvpn.ts]: Can you move the 5s to a named constant please

File: src/utils/openvpn.ts (hunk 2 above)
Comment (wrboyce) [reply src/utils/openvpn.ts]: Yeah, if the signal was _not_ successfully sent (i.e., the process does not exist, it has been killed and our work here is done).

Comment (Page-) [reply src/utils/openvpn.ts]: Ok, can you add that as a comment please.  Also it doesn't cover cases where the issue is in `kill` itself, but that should be pretty rare and we'll just fail later and restart the work/retry the kill

Comment (wrboyce) [reply src/utils/openvpn.ts]: @Page- comment added, and I've also handled the (theoretical) case where a process refuses to term/kill in a timely manner.

File: src/utils/openvpn.ts [hunk 5]
```diff
@@ -267,26 +277,66 @@ export class VpnManager extends EventEmitter {
...
+
+		return new Promise((resolve, reject) => {
+			const start = Date.now();
+			const waitForDeath = (code?: number) => {
+				if (code !== 0) {
+					// the signal was unsuccessful, the pid no longer exists;
+					// the process is dead, and our work here is done.
+					return resolve();
+				}
+				const now = Date.now();
```
Comment (Page-) [inline src/utils/openvpn.ts]: You could make this
```suggestion
				const elapsedTime = Date.now() - start;
```
and then use
```ts
if (elapsedTime  > TERM_TIMEOUT) {
```
which is a little more readable imo

Comment (wrboyce) [reply src/utils/openvpn.ts]: Agreed, will change to this. Cheers.

**Pull request merged**